File: backend/src/main/java/com/NBE3_4_2_Team4/global/security/oauth2/CustomOAuth2AccessTokenResponseClient.java
package com.NBE3_4_2_Team4.global.security.oauth2;

import lombok.extern.slf4j.Slf4j;
import org.springframework.security.oauth2.client.endpoint.*;
import org.springframework.security.oauth2.core.OAuth2RefreshToken;
import org.springframework.security.oauth2.core.endpoint.OAuth2AccessTokenResponse;
import org.springframework.stereotype.Component;
import java.util.HashMap;
import java.util.Map;
import org.springframework.security.oauth2.client.endpoint.OAuth2AccessTokenResponseClient;
import org.springframework.security.oauth2.client.endpoint.OAuth2AuthorizationCodeGrantRequest;


@Slf4j
@Component
public class CustomOAuth2AccessTokenResponseClient implements OAuth2AccessTokenResponseClient<OAuth2AuthorizationCodeGrantRequest> {
    private final RestClientAuthorizationCodeTokenResponseClient tokenResponseClient;

    public CustomOAuth2AccessTokenResponseClient() {
        this.tokenResponseClient = new RestClientAuthorizationCodeTokenResponseClient();
    }

    @Override
    public OAuth2AccessTokenResponse getTokenResponse(OAuth2AuthorizationCodeGrantRequest authorizationGrantRequest) {
        OAuth2AccessTokenResponse tokenResponse =  tokenResponseClient.getTokenResponse(authorizationGrantRequest);

        OAuth2RefreshToken refreshToken = tokenResponse.getRefreshToken();
        if (refreshToken == null) {
            log.info("❌ 리프레시 토큰 값을 받지 못했습니다. 리프레시 토큰을 지급하지 않는 서비스일 수 있습니다. 현재 서비스 : {}", authorizationGrantRequest.getClientRegistration().getClientName());
        }

        String refreshTokenValue = refreshToken == null ? null : refreshToken.getTokenValue();
        Map<String, Object> additionalParameters = new HashMap<>(tokenResponse.getAdditionalParameters());

        if (refreshTokenValue != null) {
            additionalParameters.put("refresh_token", refreshTokenValue);
        }

        return OAuth2AccessTokenResponse.withToken(
                        tokenResponse.getAccessToken().getTokenValue())
                .tokenType(tokenResponse.getAccessToken().getTokenType())
                .expiresIn(tokenResponse.getAccessToken().getExpiresAt().toEpochMilli())
                .scopes(tokenResponse.getAccessToken().getScopes())
                .refreshToken(String.valueOf(tokenResponse.getRefreshToken())) // 원래 refresh_token 값 유지
                .additionalParameters(additionalParameters) // refresh_token 포함된 추가 파라미터
                .build();
    }
}

File: backend/src/main/java/com/NBE3_4_2_Team4/global/security/oauth2/CustomOAuth2AccessTokenResponseClient.kt
package com.NBE3_4_2_Team4.global.security.oauth2

import org.slf4j.LoggerFactory
import org.springframework.security.oauth2.client.endpoint.OAuth2AccessTokenResponseClient
import org.springframework.security.oauth2.client.endpoint.OAuth2AuthorizationCodeGrantRequest
import org.springframework.security.oauth2.client.endpoint.RestClientAuthorizationCodeTokenResponseClient
import org.springframework.security.oauth2.core.endpoint.OAuth2AccessTokenResponse
import org.springframework.stereotype.Component

@Component
class CustomOAuth2AccessTokenResponseClient(
): OAuth2AccessTokenResponseClient<OAuth2AuthorizationCodeGrantRequest>{
    private val tokenResponseClient: RestClientAuthorizationCodeTokenResponseClient
    = RestClientAuthorizationCodeTokenResponseClient()

    private val log = LoggerFactory.getLogger(CustomOAuth2AccessTokenResponseClient::class.java)

    override fun getTokenResponse(authorizationGrantRequest: OAuth2AuthorizationCodeGrantRequest?): OAuth2AccessTokenResponse {
        val tokenResponse = tokenResponseClient.getTokenResponse(authorizationGrantRequest)

        val refreshToken = tokenResponse.refreshToken
        if (refreshToken == null) {
            log.info(
                "❌ 리프레시 토큰 값을 받지 못했습니다. 리프레시 토큰을 지급하지 않는 서비스일 수 있습니다. 현재 서비스 : {}",
                authorizationGrantRequest!!.clientRegistration.clientName
            )
        }

        val refreshTokenValue = refreshToken?.tokenValue
        val additionalParameters: MutableMap<String, Any> = HashMap(tokenResponse.additionalParameters)

        if (refreshTokenValue != null) {
            additionalParameters["refresh_token"] = refreshTokenValue
        }

        return OAuth2AccessTokenResponse.withToken(
            tokenResponse.accessToken.tokenValue
        )
            .tokenType(tokenResponse.accessToken.tokenType)
            .expiresIn(tokenResponse.accessToken.expiresAt!!.toEpochMilli())
            .scopes(tokenResponse.accessToken.scopes)
            .refreshToken(tokenResponse.refreshToken.toString()) // 원래 refresh_token 값 유지
            .additionalParameters(additionalParameters) // refresh_token 포함된 추가 파라미터
            .build()
    }
}